### ISSUE TITLE
Detect potentially mis-written out param functions

### ISSUE THREAD

**Issue by Walther** [rust-lang/rust-clippy#11658]
### What it does

(Feel free to adjust the title, I may have worded it wrong).

There is a significant difference between writing:
```rust
fn outparams(mut param: f64) {}

fn outparams(param: &mut f64) {}
```
but because both of these are valid code, it is easy to mistakenly write the wrong one.

It would be nice if `clippy` (or eventually even `rustc`) could detect cases where there's likely a mistake - for example when the function is pure and ends up doing nothing, something is likely to be amiss.

---

If this issue is accepted, I would love to learn how to contribute a fix for this 🧡 

### Advantage

This lint warning could detect mistakes that are easy to make and flag potentially-incorrect code for the users.

### Drawbacks

Risk of false positives would probably depend on the implementation method - I don't know enough about this yet.

### Example

I ran into this problem while directly translating some C code into Rust. I was confused about why I was always getting the same output regardless of the input, and it took me a surprisingly long time to find my error.

```rust
// bad code; this doesn't actually do anything!            vvv here
pub fn spectrum_apply_3x2(matrix: [f64; 6], src: [f64; 2], mut tgt: [f64; 2]) {
    tgt[0] = matrix[0] * src[0] + matrix[1] * src[1] + matrix[2];
    tgt[1] = matrix[3] * src[0] + matrix[4] * src[1] + matrix[5];
}
```

```rust
// the correct version                                          vvvv here
pub fn spectrum_apply_3x2(matrix: [f64; 6], src: [f64; 2], tgt: &mut [f64; 2]) {
    tgt[0] = matrix[0] * src[0] + matrix[1] * src[1] + matrix[2];
    tgt[1] = matrix[3] * src[0] + matrix[4] * src[1] + matrix[5];
}
```

(There were a lot of functions with out params like this in the larger context, so please don't nitpick the specific function body, or whether I should've used a library method instead, etc. This is just one of the smallest functions in the file, and serves as a good illustration of the issue)

[Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=2027cc68f47e12871f3bb34898b9510f) - currently neither `rustc` or `clippy` produce any warnings for the bad code.

**Comment by y21** [comment]
If I understand correctly, this lint would be quite similar to rustc's `unused_assignment` lint and it ought to lint the example here as well. I'm surprised that it doesn't already lint there (I've not looked at the code, but it seems like it ignores projections such as `tgt[0] =`, `foo.bar =`). I wonder if we could just fix this in rustc, or perhaps there's a reason it doesn't do this.
I do think that this lint would end up with a lot of the same logic that is in `unused_assignment`. For example, it would need to check that the binding is not read from again after the assignment to avoid false positives, which is what `unused_assignments` does.

An additional suggestion could then be added to that if the binding that is assigned to is a function parameter.
Something like:
```
warning: value assigned to `tgt` is never read
 --> src/main.rs:9:5
  |
9 |     tgt[0] = matrix[0] * src[0] + matrix[1] * src[1] + matrix[2];
  |     ^
  |
  = help: maybe it is overwritten before being read?
  = note: `#[warn(unused_assignments)]` on by default

help: if you meant to write to an out-param, consider changing the type of `tgt` to `&mut [f64; 2]` and dereference it
  |
8 -     pub fn spectrum_apply_3x2(matrix: [f64; 6], src: [f64; 2], mut tgt: [f64; 2]) {
8 +     pub fn spectrum_apply_3x2(matrix: [f64; 6], src: [f64; 2], tgt: &mut [f64; 2]) {

9 -      tgt[0] = matrix[0] * src[0] + matrix[1] * src[1] + matrix[2];
9 +     *tgt[0] = matrix[0] * src[0] + matrix[1] * src[1] + matrix[2];
  |
```
